protocol.md

```diff
@@ -238,6 +239,37 @@ interface CancelParams {
 
 A request that got canceled still needs to return from the server and send a response back. It can not be left open / hanging. This is in line with the JSON RPC protocol that requires that every request sends a response back. In addition it allows for returning partial results on cancel.
 
+#### <a name="partialResult"></a> Streaming Support
+
+When a language server or client is able to provide partial information before having the final result,
+it can send `$/partialResult` notifications to return parts of the result.
+
+_Notification_:
+* method: '$/partialResult'
+* params: `PartialResultParams` defined as follows:
+```typescript
+interface PartialResultParams {
+	/**
+	 * The request id to provide parts of the result for
+	 */
+	id: number | string;
+
+	/**
+	 * A JSON Patch that represents updates to the partial result as specified in RFC6902
+	 * https://tools.ietf.org/html/rfc6902
+	 */
+	path: JSONPatch;
+}
+```
+
+The receiver can listen for partial result notifications with the request ID and use the provided JSON patches to build up a result.
```

Comment:
It seems like the client should opt-in to receiving these partial results (otherwise it is wasteful). Can we add an optional (or default false) parameter to `ReferenceContext` to request streaming.

If this is opt-in, then we can also define the behavior of textDocument/references to return an empty Location array when this is true (to avoid sending the data twice).